File: app-geekhome/src/main/java/eu/geekhome/rest/configurable/ConfigurableDtoMapper.java
package eu.geekhome.rest.configurable;

import eu.geekhome.services.configurable.*;
import eu.geekhome.rest.MappingException;
import eu.geekhome.services.localization.Resource;

import javax.inject.Inject;
import java.util.List;
import java.util.stream.Collectors;

public class ConfigurableDtoMapper {

    @Inject
    private FieldDefinitionDtoMapper _fieldDefinitionDtoMapper;

    public ConfigurableDto map(Configurable configurable) throws MappingException {
        List<FieldDto> fields = null;
        Resource addNewRes = null;
        Resource editRes = null;
        if (configurable instanceof ConfigurableWithFields) {
            ConfigurableWithFields configurableWithFields = (ConfigurableWithFields)configurable;
            fields = configurableWithFields
                    .getFieldDefinitions()
                    .values()
                    .stream()
                    .map(_fieldDefinitionDtoMapper::map)
                    .collect(Collectors.toList());

            addNewRes = configurableWithFields.getAddNewRes();
            editRes = configurableWithFields.getEditRes();
        }

        return new ConfigurableDto(configurable.getTitleRes(),
                configurable.getDescriptionRes(),
                configurable.getClass().getName(),
                configurable.getParent() != null ? configurable.getParent().getName() : null,
                getConfigurableType(configurable),
                fields,
                addNewRes,
                editRes,
                configurable.getIconRaw(),
                configurable.getHasAutomation(),
                configurable.getEditableIcon(),
                configurable.getTaggable());
    }

    private ConfigurableType getConfigurableType(Configurable configurable) {
        if (configurable instanceof ConditionConfigurable) {
            return ConfigurableType.Condition;
        }

        if (configurable instanceof StateDeviceConfigurable) {
            return ConfigurableType.StateDevice;
        }

        return ConfigurableType.Other;
    }
}

File: app-geekhome/src/main/java/eu/geekhome/rest/configurable/ConfigurableDtoMapper.kt
package eu.geekhome.rest.configurable

import eu.geekhome.rest.MappingException
import eu.geekhome.rest.field.FieldDefinitionDtoMapper
import eu.geekhome.services.configurable.*
import eu.geekhome.services.localization.Resource
import java.util.stream.Collectors
import javax.inject.Inject

class ConfigurableDtoMapper @Inject constructor(
    private val fieldDefinitionDtoMapper: FieldDefinitionDtoMapper
) {
    @Throws(MappingException::class)
    fun map(configurable: Configurable): ConfigurableDto {
        var fields: List<FieldDefinitionDto>? = null
        var addNewRes: Resource? = null
        var editRes: Resource? = null
        if (configurable is ConfigurableWithFields) {
            fields = configurable
                .fieldDefinitions
                .values
                .stream()
                .map { field: FieldDefinition<*> -> fieldDefinitionDtoMapper.map(field) }
                .collect(Collectors.toList())
            addNewRes = configurable.addNewRes
            editRes = configurable.editRes
        }

        return ConfigurableDto(
            configurable.titleRes,
            configurable.descriptionRes,
            configurable.javaClass.name,
            if (configurable.parent != null) configurable.parent!!.name else null,
            getConfigurableType(configurable),
            fields,
            addNewRes,
            editRes,
            configurable.iconRaw,
            configurable.hasAutomation,
            configurable.editableIcon,
            configurable.taggable
        )
    }

    private fun getConfigurableType(configurable: Configurable): ConfigurableType {
        if (configurable is ConditionConfigurable) {
            return ConfigurableType.Condition
        }
        return if (configurable is StateDeviceConfigurable) {
            ConfigurableType.StateDevice
        } else ConfigurableType.Other
    }
}